File: android/app/src/main/java/com/noop/analytics/PhysiologicalSteps.kt
package com.noop.analytics

/**
 * WHOOP-style step-day boundaries. A physiological step day starts at the onset of the main sleep
 * and ends at the next main-sleep onset; the delta is still attributed to the later counter sample.
 * We deliberately do not mask the sleep span: a real walk during a long night-time wake remains a
 * walk, while the shared activity-class and tick-rate gates continue to reject non-locomotion.
 */
object PhysiologicalSteps {

    enum class SleepKind { MAIN_SLEEP, NAP, UNCLASSIFIED }

    data class SleepBlock(
        val onset: Long,
        val end: Long,
        /** Stable identity lets a user-edited onset move an existing boundary instead of adding one. */
        val id: String = onset.toString(),
        val editedOnset: Long? = null,
        val kind: SleepKind = SleepKind.UNCLASSIFIED,
    ) {
        val effectiveOnset: Long get() = editedOnset ?: onset
    }

    /** The one open physiological cycle. Absence of a newly detected night never creates a calendar cycle. */
    data class CycleBoundary(val sleepId: String, val onset: Long)

    data class CycleWindow(val sleepId: String, val onset: Long, val endExclusive: Long)

    data class OwnerSegment(val owner: String, val onset: Long, val endExclusive: Long)
    data class OwnerCoverage(
        val owner: String,
        val onset: Long,
        val endExclusive: Long,
        val priority: Int,
    )

    fun shouldReadCounterPredecessor(segmentIndex: Int): Boolean = segmentIndex == 0

    private const val MIN_MAIN_SLEEP_SECONDS = 3 * 3_600L


    /**
     * Apply the same user-visible main-vs-nap shape used by the sleep surfaces: only the canonical overnight
     * group can be MAIN, it must total at least three hours, and every other/twinless-explicit block is NAP.
     */
    fun classifyForCycle(
        blocks: List<SleepBlock>,
        tzOffsetSeconds: Long,
        habitualMidsleepSec: Long?,
    ): List<SleepBlock> {
        if (blocks.isEmpty()) return emptyList()
        val explicitMain = blocks.indices.filter { blocks[it].kind == SleepKind.MAIN_SLEEP }
        val selectable = blocks.indices.filter { blocks[it].kind != SleepKind.NAP }
        val selectedOriginalIndices = if (explicitMain.isNotEmpty()) {
            explicitMain
        } else {
            val selectableNightBlocks = selectable.map {
                val b = blocks[it]
                SleepStageTotals.NightBlock(b.effectiveOnset, b.end)
            }
            // Eliminate nap-shaped GROUPS before choosing a winner. Otherwise a six-hour afternoon nap can
            // win the generic duration scorer, fail the daytime guard, and hide a valid shorter night.
            val eligible = SleepStageTotals.bridgedNightGroups(selectableNightBlocks, tzOffsetSeconds)
                .filter { group ->
                    val total = group.indices.sumOf { i -> selectableNightBlocks[i].durationS.coerceAtLeast(0L) }
                    val onset = group.indices.minOfOrNull { selectableNightBlocks[it].start }
                    total >= MIN_MAIN_SLEEP_SECONDS && onset != null &&
                        SleepStageTotals.isOvernightOnset(onset, tzOffsetSeconds)
                }
                .flatMap { it.indices }
                .distinct()
            SleepStageTotals.mainNightGroupIndices(
                eligible.map { selectableIndex ->
                    val original = selectable[selectableIndex]
                    val b = blocks[original]
                    SleepStageTotals.NightBlock(b.effectiveOnset, b.end)
                },
                tzOffsetSeconds,
                habitualMidsleepSec,
            ).orEmpty().map { selectedIndex ->
                selectable[eligible[selectedIndex]]
            }
        }.toHashSet()
        return blocks.mapIndexed { index, block ->
            block.copy(kind = if (index in selectedOriginalIndices) SleepKind.MAIN_SLEEP else SleepKind.NAP)
        }
    }

    /** Use the exact same main-night/group selector as the Sleep screen and daily sleep aggregate. */
    fun mainSleepOnset(
        blocks: List<SleepBlock>,
        tzOffsetSeconds: Long,
        habitualMidsleepSec: Long?,
    ): Long? {
        val eligible = classifyForCycle(blocks, tzOffsetSeconds, habitualMidsleepSec)
            .filter { it.kind == SleepKind.MAIN_SLEEP }
        val indices = SleepStageTotals.mainNightGroupIndices(
            eligible.map { SleepStageTotals.NightBlock(it.effectiveOnset, it.end) },
            tzOffsetSeconds,
            habitualMidsleepSec,
        ) ?: return null
        return indices.minOfOrNull { eligible[it].effectiveOnset }
    }

    /**
     * Build historical sleep-to-sleep windows without inventing rows at midnight. The final observed
     * boundary alone owns the open tail to [now], including days on which no new sleep was detected.
     */
    fun cycleWindows(boundaries: List<CycleBoundary>, now: Long): List<CycleWindow> {
        val ordered = boundaries.asSequence()
            .filter { it.onset <= now }
            .distinctBy { it.sleepId }
            .sortedBy { it.onset }
            .toList()
        return ordered.mapIndexedNotNull { index, boundary ->
            val end = ordered.getOrNull(index + 1)?.onset ?: now
            if (end <= boundary.onset) null else CycleWindow(boundary.sleepId, boundary.onset, end)
        }
    }

    /** Split at local midnights and merge adjacent same-owner slices; counters never cross device seams. */
    fun ownerSegments(
        window: CycleWindow,
        ownerByDay: Map<String, String>,
        tzOffsetSeconds: Long,
        fallbackOwner: String,
    ): List<OwnerSegment> {
        if (window.endExclusive <= window.onset) return emptyList()
        val out = ArrayList<OwnerSegment>()
        var cursor = window.onset
        var lastOwner = fallbackOwner
        while (cursor < window.endExclusive) {
            val day = AnalyticsEngine.dayString(cursor, tzOffsetSeconds)
            val local = cursor + tzOffsetSeconds
            val dayNumber = Math.floorDiv(local, SleepStageTotals.SECONDS_PER_DAY)
            val dayStart = dayNumber * SleepStageTotals.SECONDS_PER_DAY - tzOffsetSeconds
            // The sleep-owner owns the partial onset day. Calendar owner changes take effect only at a
            // midnight seam, never retroactively before the boundary sleep began.
            val owner = if (cursor == window.onset && cursor > dayStart) {
                fallbackOwner
            } else {
                ownerByDay[day] ?: lastOwner
            }
            val nextMidnight = (dayNumber + 1) * SleepStageTotals.SECONDS_PER_DAY - tzOffsetSeconds
            val end = minOf(window.endExclusive, nextMidnight)
            val previous = out.lastOrNull()
            if (previous != null && previous.owner == owner && previous.endExclusive == cursor) {
                out[out.lastIndex] = previous.copy(endExclusive = end)
            } else {
                out += OwnerSegment(owner, cursor, end)
            }
            lastOwner = owner
            cursor = end
        }
        return out
    }

    /**
     * Timestamp coverage wins at every seam; priority resolves overlap, so two worn straps never add.
     * Each coverage is a continuous first..last ownership span by design: an internal sample/radio gap
     * creates no seam and therefore cannot hand the interval to a lower-priority concurrently worn strap.
     */
    fun ownerSegmentsFromCoverage(
        window: CycleWindow,
        coverage: List<OwnerCoverage>,
        fallbackOwner: String,
    ): List<OwnerSegment> {
        if (window.endExclusive <= window.onset) return emptyList()
        val clipped = coverage.mapNotNull {
            val start = maxOf(window.onset, it.onset)
            val end = minOf(window.endExclusive, it.endExclusive)
            if (end <= start) null else it.copy(onset = start, endExclusive = end)
        }
        val seams = buildSet {
            add(window.onset)
            add(window.endExclusive)
            clipped.forEach { add(it.onset); add(it.endExclusive) }
        }.sorted()
        val out = ArrayList<OwnerSegment>()
        var lastOwner = fallbackOwner
        for (i in 0 until seams.lastIndex) {
            val start = seams[i]
            val end = seams[i + 1]
            val owner = clipped.asSequence()
                .filter { start >= it.onset && start < it.endExclusive }
                .minWithOrNull(compareBy<OwnerCoverage> { it.priority }.thenBy { it.owner })
                ?.owner ?: lastOwner
            val previous = out.lastOrNull()
            if (previous != null && previous.owner == owner && previous.endExclusive == start) {
                out[out.lastIndex] = previous.copy(endExclusive = end)
            } else {
                out += OwnerSegment(owner, start, end)
            }
            lastOwner = owner
        }
        return out
    }

    /** Resolve the next observed onset, or [now] for the currently open physiological cycle. */
    fun cycleEnd(day: String, onsetByDay: Map<String, Long>, now: Long): Long {
        val onset = onsetByDay[day] ?: return now
        return onsetByDay.values.asSequence()
            .filter { it > onset && it <= now }
            .minOrNull()
            ?: now
    }

}

File: Packages/StrandAnalytics/Sources/StrandAnalytics/PhysiologicalSteps.swift
import Foundation
import WhoopProtocol

public enum PhysiologicalSteps {
    public enum SleepKind: Sendable { case mainSleep, nap, unclassified }
    public struct SleepBlock: Sendable {
        public let onset: Int
        public let end: Int
        public let id: String
        public let editedOnset: Int?
        public let kind: SleepKind
        public var effectiveOnset: Int { editedOnset ?? onset }
        public init(onset: Int, end: Int, id: String? = nil, editedOnset: Int? = nil,
                    kind: SleepKind = .unclassified) {
            self.onset = onset; self.end = end; self.id = id ?? String(onset)
            self.editedOnset = editedOnset; self.kind = kind
        }
    }
    public struct CycleBoundary: Equatable, Sendable {
        public let sleepId: String
        public let onset: Int
        public init(sleepId: String, onset: Int) { self.sleepId = sleepId; self.onset = onset }
    }
    public struct CycleWindow: Equatable, Sendable {
        public let sleepId: String; public let onset: Int; public let endExclusive: Int
        public init(sleepId: String, onset: Int, endExclusive: Int) {
            self.sleepId = sleepId; self.onset = onset; self.endExclusive = endExclusive
        }
    }
    public struct OwnerSegment: Equatable, Sendable {
        public let owner: String; public let onset: Int; public let endExclusive: Int
        public init(owner: String, onset: Int, endExclusive: Int) {
            self.owner = owner; self.onset = onset; self.endExclusive = endExclusive
        }
    }
    public struct OwnerCoverage: Equatable, Sendable {
        public let owner: String; public let onset: Int; public let endExclusive: Int; public let priority: Int
        public init(owner: String, onset: Int, endExclusive: Int, priority: Int) {
            self.owner = owner; self.onset = onset; self.endExclusive = endExclusive; self.priority = priority
        }
    }
    private static let minMainSleepSeconds = 3 * 3_600

    /// Kotlin twin: `PhysiologicalSteps.classifyForCycle`.
    public static func classifyForCycle(_ blocks: [SleepBlock], offsetSec: Int,
                                        habitualMidsleepSec: Int?) -> [SleepBlock] {
        guard !blocks.isEmpty else { return [] }
        let explicit = blocks.indices.filter { blocks[$0].kind == .mainSleep }
        let selectable = blocks.indices.filter { blocks[$0].kind != .nap }
        let selected: Set<Int>
        if !explicit.isEmpty {
            selected = Set(explicit)
        } else {
            let nightBlocks = selectable.map { SleepStageTotals.NightBlock(start: blocks[$0].effectiveOnset,
                                                                            end: blocks[$0].end) }
            let eligible = SleepStageTotals.bridgedNightGroups(nightBlocks, offsetSec: offsetSec)
                .filter { group in
                    let total = group.indices.reduce(0) { $0 + max(0, nightBlocks[$1].durationS) }
                    let onset = group.indices.map { nightBlocks[$0].start }.min()
                    return total >= minMainSleepSeconds && onset.map {
                        SleepStageTotals.isOvernightOnset($0, offsetSec: offsetSec)
                    } == true
                }
                .flatMap { $0.indices }
            let candidates = eligible.map { index in
                SleepStageTotals.NightBlock(start: blocks[selectable[index]].effectiveOnset,
                                            end: blocks[selectable[index]].end)
            }
            let picked = SleepStageTotals.mainNightGroupIndices(candidates, offsetSec: offsetSec,
                                                                habitualMidsleepSec: habitualMidsleepSec) ?? []
            selected = Set(picked.map { selectable[eligible[$0]] })
        }
        return blocks.indices.map { index in
            let block = blocks[index]
            return SleepBlock(onset: block.onset, end: block.end, id: block.id,
                              editedOnset: block.editedOnset,
                              kind: selected.contains(index) ? .mainSleep : .nap)
        }
    }

    /// Kotlin twin: `PhysiologicalSteps.cycleWindows`.
    public static func cycleWindows(_ boundaries: [CycleBoundary], now: Int) -> [CycleWindow] {
        var ids = Set<String>()
        let ordered = boundaries.filter { $0.onset <= now && ids.insert($0.sleepId).inserted }
            .sorted { $0.onset < $1.onset }
        return ordered.indices.compactMap { index in
            let end = index + 1 < ordered.count ? ordered[index + 1].onset : now
            return end > ordered[index].onset
                ? CycleWindow(sleepId: ordered[index].sleepId, onset: ordered[index].onset, endExclusive: end)
                : nil
        }
    }

    /// Kotlin twin: `PhysiologicalSteps.ownerSegmentsFromCoverage`.
    public static func ownerSegmentsFromCoverage(_ window: CycleWindow, coverage: [OwnerCoverage],
                                                 fallbackOwner: String) -> [OwnerSegment] {
        guard window.endExclusive > window.onset else { return [] }
        let clipped = coverage.compactMap { item -> OwnerCoverage? in
            let start = max(window.onset, item.onset), end = min(window.endExclusive, item.endExclusive)
            return end > start ? OwnerCoverage(owner: item.owner, onset: start,
                endExclusive: end, priority: item.priority) : nil
        }
        let seams = Set([window.onset, window.endExclusive] + clipped.flatMap { [$0.onset, $0.endExclusive] }).sorted()
        var result: [OwnerSegment] = [], lastOwner = fallbackOwner
        for index in 0..<(seams.count - 1) {
            let start = seams[index], end = seams[index + 1]
            let owner = clipped.filter { start >= $0.onset && start < $0.endExclusive }
                .sorted { $0.priority == $1.priority ? $0.owner < $1.owner : $0.priority < $1.priority }
                .first?.owner ?? lastOwner
            if let last = result.last, last.owner == owner, last.endExclusive == start {
                result[result.count - 1] = OwnerSegment(owner: owner, onset: last.onset, endExclusive: end)
            } else {
                result.append(OwnerSegment(owner: owner, onset: start, endExclusive: end))
            }
            lastOwner = owner
        }
        return result
    }

}
